### PR TITLE
implement open service

### PR DESCRIPTION
**Pull request by MaaooVaass**
## Summary
implement open service and create new service folder

## Card
https://trello.com/c/YQlP8f2F/36-integraci%C3%B3n-api-externa

### CODE REVIEW

File: app/controllers/books_controller.rb [hunk 1]
```diff
@@ -10,4 +10,15 @@ def show
   def index
     render_paginated Book.all
   end
+
+  def info_book
+    isbn = params[:isbn]
```
Comment (lmasello) [inline app/controllers/books_controller.rb]: Redundant variable, you can just use the `params[:isbn]`

File: app/controllers/books_controller.rb [hunk 2]
```diff
@@ -10,4 +10,15 @@ def show
...
+    return render json: { error: t('.no_isbn') }, status: :bad_request if isbn.nil?
```
Comment (lmasello) [inline app/controllers/books_controller.rb]: Where is this translation defined?

Comment (lmasello) [reply app/controllers/books_controller.rb]: Also, you can use strong parameters and avoid checking that the parameter is being sent.
Some method similar to this:
```ruby
def book_info_params
  params.require(:isbn)
end
```

File: app/controllers/books_controller.rb [hunk 3]
```diff
@@ -10,4 +10,15 @@ def show
...
+
+    book_info = OpenLibraryService.new(isbn).book_info
+    return render json: { error: t('.no_book_error', code: isbn) }, status: :not_found if
```
Comment (lmasello) [inline app/controllers/books_controller.rb]: No need to check whether the external service returns nothing or not.

File: app/controllers/books_controller.rb [hunk 4]
```diff
@@ -10,4 +10,15 @@ def show
...
+    book_info.empty?
+
```
Comment (lmasello) [inline app/controllers/books_controller.rb:24]: Empty line :)

File: app/controllers/books_controller.rb (hunk 3 above)
Comment (fedeest) [inline app/controllers/books_controller.rb]: Don't do a one line if it doesn't fit in one line.

File: app/controllers/books_controller.rb (hunk 2 above)
Comment (fedeest) [inline app/controllers/books_controller.rb]: Where's this translation added?

File: app/controllers/books_controller.rb [hunk 5]
```diff
@@ -10,4 +10,15 @@ def show
...
+    render json: book_info
+  end
```
Comment (fedeest) [inline app/controllers/books_controller.rb:27]: Add tests for this endpoint

**Review by fedeest — CHANGES_REQUESTED** [review]

**Review by lmasello — CHANGES_REQUESTED** [review]

File: app/services/open_library_service.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        format: FORMAT,
+        jscmd: JSCMD
+      }
+    }
+  end
+
+  def book_info
+    response = retrieve_from_api
+    return {} if response.nil?
+
```
Comment (lmasello) [inline app/services/open_library_service.rb:22]: Empty line

File: app/services/open_library_service.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+class OpenLibraryService
```
Comment (lmasello) [inline app/services/open_library_service.rb:1]: We should also test this Service with rspec. For that purposes we need to mock the response of the external API. I recommend that you read the following article: https://8thlight.com/blog/dariusz-pasciak/2016/09/08/better-rspec-stubbing.html

We can make a call after that

**Review by lmasello — CHANGES_REQUESTED** [review]

File: app/controllers/books_controller.rb [hunk 8]
```diff
@@ -4,11 +4,20 @@ class BooksController < ApiController
...
   end
 
   def index
-    @books = Book.all
-    render_paginated @books
+    render_paginated Book.all
+  end
+
+  def info_book
+    params.require(:isbn)
```
Comment (lmasello) [inline app/controllers/books_controller.rb]: You can create a private method for this param, such as `info_book_params` and then in the OpenLibraryService call, do `OpenLibraryService.new(info_book_params).book_info`

File: config/locales/en.yml [hunk 9]
```diff
@@ -1,8 +1,10 @@
...
-  current_date: "Current date: %{current_date}"
-  book_title: "Book title: %{book_title}"
-  book_genre: "Book genre: %{book_genre}"
-  book_author: "Book author: %{book_author}"
\ No newline at end of file
+  greeting: "Hello: %{name}"
+  start_date: "Start date: %{star_d}"
+  end_date: "End date: %{end_d}"
+  current_date: "Current date: %{current_d}"
+  book_title: "Book title: %{book_t}"
```
Comment (lmasello) [inline config/locales/en.yml]: Avoid unnecessary abbreviations

**Review by lmasello — CHANGES_REQUESTED** [review]

File: spec/services/open_library_spec.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                                              { 'url' => 'https://openlibrary.org/subjects/',
+                                                'name' => 'Zen Buddhism' }],
+                               'publish_date' => '1994',
+                               'key' => '/books/OL1397864M',
+                               'authors' => [{ 'url' => 'https://openlibrary.org/autho/Zhizhg_Cai',
+                                               'name' => 'Zhizhong Cai' }],
+                               'classifications' => { 'dewey_decimal_class' => ['294.3/927'],
+                                                      'lc_classifications' => ['BQ926.T7311994'] },
+                               'publish_places' => [{ 'name' => 'New York' }] }
+      }.to_json
```
Comment (fedeest) [inline spec/services/open_library_spec.rb]: Somethings not right here. You are missing the `end` from the `do` block of the `let`

File: spec/services/open_library_spec.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+                                                 "title": 'Zen speaks',
+                                                 "subtitle": 'shouts of nothingness',
+                                                 "number_of_pages": 159,
+                                                 "authors": ['Zhizhong Cai'])
+      end
+
+      context 'When fetching a book info that not exist' do
+        let(:isbn) { -1 }
+        let(:hash_book_info) { {} }
+        it 'responds with 404 if the book info do not exist' do
```
Comment (fedeest) [inline spec/services/open_library_spec.rb]: Change this description accordingly

**Review by fedeest — CHANGES_REQUESTED** [review]

**Review by fedeest — APPROVED** [review]

**Review by lmasello — APPROVED** [review]
Well done! :+1:

**Pull request merged**